rfc/src/rfcs/0003-resource-limits-timeouts.md

```diff
@@ -0,0 +1,224 @@
+
+> **Question 2: Should we include units in the attributes?**
+>
+> I'm presently giving "seconds" and "gigabytes" and these seem totally reasonable units that we probably wouldn't ever deviate from.
+> However, it might be nice to either require units (to be explicit) or include units (if we think people might need them... perhaps just minutes).
+> I believe it would be reasonable to omit units however, and I'm not sure how difficult it would be to support `timeout = 2min` or `timeout = 4:00`.
+
+Beyond specifying limits on each proof harness, users would interact with these limits in two other ways:
+
+1. The output from running Kani.
+2. The command line arguments for manipulating limits.
+
+> **Question 3: How much flexibility should be provide on the command-line?**
+>
+> I believe we can keep this small.
+> I propose we only add a few new options, and I'm not sure if we should even go that far:
+> 1. `--default-timeout`
+> 2. `--default-memory`
+> 3. `--timeout-multiplier=X`
+> 4. `--ignore-limits`
+>
+> The purpose of the first two is not really for the command line but instead `Cargo.toml` (which currently is just an alternative way to supply command line arguments).
+>
+> The purpose of the multiplier is to allow a slow/inconsistent CI system to have an extra "grace period" for timeouts.
+> This way, CI can run `cargo kani --timeout-multiplier=1.3` to get 30% extra slack on timeouts.
+>
+> The `--ignore-limits` options should accomplish two things:
+> 1. Replicate current behavior.
+> 2. Allow Kani to measure memory usage and CPU time, and report suggested limits.
+
+For output, we have a number of differences from current Kani:
+
+1. When timeout triggers during unwinding, we can report `unwind` as a suggestion.
+2. When timeout triggers during solving, we can suggest raising the `timeout` or reducing the problem size.
+3. When memory limits get hit, we can suggest raising the limit. (Or: reporting a Kani bug perhaps?)
+4. If verification completes during `--timeout-multiplier` "grace period" we can emit a warning.
+5. If limits were exceeded but with `--ignore-limits`, we can report that actual CPU/memory usage.
+
+> **Question 4: Should we emit a warning when verification time comes close to (but under) the timeout?**
+> Ditto memory usage. I'm not sure.
+
+Our tutorial and documentation will need updating as a result of integrating this feature.
+Parts of the tutorial can be improved (such as the unwinding section) as we no longer need `unwind(1)` to force termination.
+We might wish to write specific documentation to link to from (e.g.) the error message we emit when a limit was exceeded.
+
+## Detailed Design
+
+This will impact:
+
+1. Two new attributes in the `kani` library: `kani::timeout(X)` and `kani::memory(X)`.
+Also `kani::proof` if we decide to add convenient annotations there.
+2. `kani-compiler` and `kani_metadata` will be updated to record this metadata.
+3. `kani-driver` needs to updated to apply these limits.
```

Comment:
How is `kani-driver` going to enforce those limits? How will this evolve if we start using CBMC as a library instead of a separate process?